app/src/org/commcare/engine/cases/CaseIndexQuerySetTransform.java

```diff
@@ -0,0 +1,86 @@
+package org.commcare.engine.cases;
+
+import org.commcare.cases.model.CaseIndex;
+import org.commcare.cases.query.QueryContext;
+import org.commcare.cases.query.queryset.CaseModelQuerySetMatcher;
+import org.commcare.cases.query.queryset.CaseQuerySetLookup;
+import org.commcare.cases.query.queryset.DerivedCaseQueryLookup;
+import org.commcare.cases.query.queryset.DualTableSingleMatchModelQuerySet;
+import org.commcare.cases.query.queryset.ModelQuerySet;
+import org.commcare.cases.query.queryset.QuerySetLookup;
+import org.commcare.cases.query.queryset.QuerySetTransform;
+import org.commcare.models.database.user.models.CaseIndexTable;
+import org.javarosa.core.model.instance.TreeReference;
+import org.javarosa.core.model.trace.EvaluationTrace;
+
+import java.util.Set;
+
+/**
+ * Created by ctsims on 2/6/2017.
+ */
+
+public class CaseIndexQuerySetTransform implements QuerySetTransform {
+
+    private final CaseIndexTable table;
+
+    public CaseIndexQuerySetTransform(CaseIndexTable table) {
+        this.table = table;
+    }
+
+    @Override
+    public QuerySetLookup getTransformedLookup(QuerySetLookup incoming, TreeReference relativeLookup) {
+        if(incoming.getQueryModelId().equals(CaseQuerySetLookup.CASE_MODEL_ID)) {
+            if(relativeLookup.size() == 2 && "index".equals(relativeLookup.getName(0))) {
+                String indexName = relativeLookup.getName(1);
+                return new CaseIndexQuerySetLookup(indexName, incoming, table);
+            }
+        }
+        return null;
+    }
+
+
+
+    public static class CaseIndexQuerySetLookup extends DerivedCaseQueryLookup {
+        String indexName;
+        CaseIndexTable table;
```

Comment:
Hmm, looks like you only use the bulkReadIndexToCaseIdMatch of this class; if we put that in an interface we could move this class into commcare-core